From 49836d39f28b1cdff80d0e51a07fa106b596a93a Mon Sep 17 00:00:00 2001
From: lilydjwg <lilydjwg@gmail.com>
Date: Fri, 19 Jan 2024 18:16:45 +0800
Subject: [PATCH] input method: don't send a release event without a
 corresponding press event to IM

e.g. in Firefox, place focus anywhere not a text input, press Ctrl-L to
focus the urlbar, l would be pressed for Firefox continuously without
this patch because we got two release events for one press event.
---
 src/core/seat/input-method-relay.cpp | 21 +++++++++++++++++++++
 src/core/seat/input-method-relay.hpp |  2 ++
 2 files changed, 23 insertions(+)

diff --git a/src/core/seat/input-method-relay.cpp b/src/core/seat/input-method-relay.cpp
index a341ed37..8cb05145 100644
--- a/src/core/seat/input-method-relay.cpp
+++ b/src/core/seat/input-method-relay.cpp
@@ -221,6 +221,22 @@ bool wf::input_method_relay::should_grab(wlr_keyboard *kbd)
     return !is_im_sent(kbd);
 }
 
+bool wf::input_method_relay::check_superfluous_release(uint32_t key, uint32_t state)
+{
+    if (state == WL_KEYBOARD_KEY_STATE_PRESSED)
+    {
+        this->pressed_keys.insert(key);
+        return false;
+    } else if (auto it = this->pressed_keys.find(key); it != this->pressed_keys.end())
+    {
+        this->pressed_keys.erase(it);
+        return false;
+    } else
+    {
+        return true;
+    }
+}
+
 bool wf::input_method_relay::is_im_sent(wlr_keyboard *kbd)
 {
     struct wlr_virtual_keyboard_v1 *virtual_keyboard = wlr_input_device_get_virtual_keyboard(&kbd->base);
@@ -261,6 +277,11 @@ bool wf::input_method_relay::handle_key(struct wlr_keyboard *kbd, uint32_t time,
         return false;
     }
 
+    if (check_superfluous_release(key, state))
+    {
+        return false;
+    }
+
     wlr_input_method_keyboard_grab_v2_set_keyboard(keyboard_grab, kbd);
     wlr_input_method_keyboard_grab_v2_send_key(keyboard_grab, time, key, state);
     return true;
diff --git a/src/core/seat/input-method-relay.hpp b/src/core/seat/input-method-relay.hpp
index 1d686475..96d4ac3b 100644
--- a/src/core/seat/input-method-relay.hpp
+++ b/src/core/seat/input-method-relay.hpp
@@ -41,6 +41,8 @@ class input_method_relay
     };
 
     bool should_grab(wlr_keyboard*);
+    std::multiset<uint32_t> pressed_keys;
+    bool check_superfluous_release(uint32_t key, uint32_t state);
 
   public:
 
-- 
2.43.0

